Adds retail & f-takes-all disk formatting

gaasedelen opened the pull request:
I wrote these additions today and am making this PR as near-complete. It will help address #18 and #30. I also don't really want to be developing on a moving target so I'd rather get discussion and feedback going now :-P

This PR will add the following flags:

```
Disk formatting options:
    --format=<format>              specify the format (retail, f-takes-all) to initialize the device to
    --sector-size=<size>           specify the size (in bytes) of a device sector (default is 512)
    --sectors-per-cluster=<size>   specify the sectors per cluster when initializing non-retail partitions (default is 128)
    --destroy-all-existing-data    acknowledge that device formatting will destroy all existing data
```

Example usage:

```
fatxfs /dev/sdb --format=retail --destroy-all-existing-data
```

Disk formatting should be identical to hardware. 

~~However, tip/master of the main FATX repo appears to have regressed something since 44a03f82b18f4a49a6a4b02361cc469718e7af76 and I rebased onto it without checking, whoops. I can't seem to write a file above a few KB to a 4096 byte sector drive anymore. Previously I was able to, and these files worked on both on Hardware and in XEMU.~~

There are a few other fixes that ended up in here. Since code is currently churning on this repo, I hope you don't mind them getting lumped into this. I'll try to leave comments on a few points of the PR that might need to be discussed further.

Please feel free to make any changes.

EDIT: By 'near-complete' I mean I haven't been able to fully test the end-to-end workflow of initializing a disk, writing a file into it, and loading it on XEMU/HW with master regressed. Also I imagine there might be some questions/changes/cleanup that may be requested 👍

gaasedelen commented on src/fatx_disk.c:
```diff
@@ -0,0 +1,153 @@
+    { .letter = 'y',    .offset = 0x2ee80000, .size = 0x02ee00000 },
+    { .letter = 'z',    .offset = 0x5dc80000, .size = 0x02ee00000 },
+    { .letter = 'c',    .offset = 0x8ca80000, .size = 0x01f400000 },
+    { .letter = 'e',    .offset = 0xabe80000, .size = 0x1312d6000 },
```
The size of the E partition is incorrect on master. It should be `0x1312d6000` not `0x131f00000`

gaasedelen commented on src/fatx_fat.c:
```diff
@@ -20,12 +20,118 @@
 static bool fatx_cluster_valid(struct fatx_fs *fs, size_t cluster)
 {
-    return (cluster >= FATX_FAT_RESERVED_ENTRIES_COUNT) &&
+    return (cluster >= 0) &&
```
This might be a controversial change which is which I am leaving a comment about it. Without this, `fatx_write_fat(...)` will fail in `fatx_init_root(...)` when trying to write 0xFFF8 to the root directory cluster id (0).

gaasedelen commented on src/fatx_internal.h:
```diff
@@ -46,7 +40,7 @@
 
 /* Mask to be applied when reading FAT entry values. */
 #define FATX_FAT16_ENTRY_MASK        0x0000ffff
-#define FATX_FAT32_ENTRY_MASK        0x0fffffff
```
`0x0ffffffff` appears to have been a FAT thing? Maybe? But I do not think it applies to FATX.

gaasedelen commented on src/fatx_partition.c:
```diff
@@ -50,15 +51,30 @@ int fatx_check_partition_signature(struct fatx_fs *fs)
+{
+    struct timeval time;
+
+    gettimeofday(&time, NULL);
+    fs->volume_id = time.tv_usec;
```
Technically the volume ID is just random, but this is what the xbox itself does (effectively) so *shrug*.

gaasedelen commented on src/fatxfs_fuse.c:
```diff
@@ -619,17 +612,22 @@ void fatx_fuse_print_usage(void)
+                    "    --format=<format>              specify the format (retail, f-takes-all) to initialize the device to\n"
+                    "    --sector-size=<size>           specify the size (in bytes) of a device sector (default is 512)\n"
+                    "    --sectors-per-cluster=<size>   specify the sectors per cluster when initializing non-retail partitions (default is 128)\n"
+                    "    --destroy-all-existing-data    acknowledge that device formatting will destroy all existing data\n\n");
```
`--destroy-all-existing-data` is simply a safety feature designed to try and force users to acknowledge the consequences of what they're trying to do. It is mandatory (currently) to include this flag if also using `--format`

gaasedelen commented on src/fatx_fat.c:
```diff
@@ -103,40 +209,36 @@ int fatx_get_fat_entry_type(struct fatx_fs *fs, fatx_fat_entry entry)
+                return FATX_CLUSTER_DATA;
+            case 0xffff:
+                return FATX_CLUSTER_END;
+        }
     }
     else
     {
         entry &= FATX_FAT32_ENTRY_MASK;
-
-        if (entry == 0)
-            return FATX_CLUSTER_AVAILABLE;
-
-        if (entry == 1)
-            return FATX_CLUSTER_RESERVED;
-
-        if (entry >= FATX_FAT_RESERVED_ENTRIES_COUNT && entry <= 0x0ffffff6)
-            return FATX_CLUSTER_DATA;
-
-        if (entry == 0x0ffffff7)
-            return FATX_CLUSTER_BAD;
-
-        if (entry >= 0x0ffffff8 && entry <= 0x0fffffff)
-            return FATX_CLUSTER_END;
+        switch (entry)
+        {
+            case 0x00000000:
+                return FATX_CLUSTER_AVAILABLE;
+            case 0xfffffff0:
+                return FATX_CLUSTER_RESERVED;
+            case 0xfffffff7:
+                return FATX_CLUSTER_BAD;
+            case 0xfffffff8:
+                return FATX_CLUSTER_DATA;
+            case 0xffffffff:
+                return FATX_CLUSTER_END;
+        }
     }
```
TBH this whole func seemed very out-of-spec? Maybe it was written based on non-xbox FAT documentation. I noticed when I was trying to write the root directory entry.

With my changes I tried to stick to the facts and simplify.

gaasedelen commented on src/fatx_internal.h:
```diff
@@ -88,10 +82,12 @@
-    uint32_t unknown1;
+    uint32_t root_cluster;
+    uint16_t unknown1;
+    uint8_t  padding[4078];
```
`root_cluster` is supposed to be a 32bit field in the superblock.

The `unknown1` is technically supposed to be a unicode volume name of some length but I've only ever seen it as a single null terminator `\x00\x00` in retail.

Maybe it is used on memory cards? But I don't think it really matters. either way.

gaasedelen commented:
The above comments are meant to call out some of the 'odd end' changes that made their way into the larger PR.

The rest of the material should probably be reviewed but will hopefully make more sense as simple additions towards formatting drives.

mborgerson commented on src/fatx_fat.c:
```diff
@@ -20,12 +20,118 @@
+    uint8_t *chunk;
+    int retval = FATX_STATUS_SUCCESS;
+
+    if (fatx_write_fat(fs, 0, 0xfffffff8) || fatx_mark_cluster_end(fs, 1))
```
Let's use `fs->root_cluster` instead of `1`

mborgerson commented on src/fatx_fat.c:
```diff
@@ -20,12 +20,118 @@
+
+    while (bytes_remaining > 0)
+    {
+        if (bytes_remaining > chunk_size)
```
Can we simplify to something like:

```c
while (bytes_remaining > 0) {
    size_t bytes_to_write = MIN(chunk_size, bytes_remaining);
    if (fatx_dev_write(fs, chunk, bytes_to_write, 1) != 1) {
        ....
    }
    bytes_remaining -= bytes_to_write;
}

mborgerson commented:
> I wrote these additions today and am making this PR as near-complete. It will help address #18 and #30. I also don't really want to be developing on a moving target so I'd rather get discussion and feedback going now :-P

Awesome. Thanks for the patch!

> 
> This PR will add the following flags:
> 
> ```
> Disk formatting options:
>     --format=<format>              specify the format (retail, f-takes-all) to initialize the device to
>     --sector-size=<size>           specify the size (in bytes) of a device sector (default is 512)
>     --sectors-per-cluster=<size>   specify the sectors per cluster when initializing non-retail partitions (default is 128)
>     --destroy-all-existing-data    acknowledge that device formatting will destroy all existing data
> ```
> 
> Example usage:
> 
> ```
> fatxfs /dev/sdb --format=retail --destroy-all-existing-data
> ```
> 
> Disk formatting should be identical to hardware.

It will not be exactly 1:1 with a retail box out of the factory because it is missing the config partition. I'm not sure if a retail box will boot without the config partition set up, a quick test in xemu would confirm; but the blank I created before simply has the refurbish data signature `BRFR` at 0x600 bytes and the rest zeroed from 0-0x80000 bytes. We might want to do the same here.

> However, tip/master of the main FATX repo appears to have regressed something since [44a03f8](https://github.com/mborgerson/fatx/commit/44a03f82b18f4a49a6a4b02361cc469718e7af76) and I rebased onto it without checking, whoops. I can't seem to write a file above a few KB to a 4096 byte sector drive anymore. Previously I was able to, and these files worked on both on Hardware and in XEMU.

Weird. The tester script writes a (~64K) file and reads it back out. You have confirmed it's an issue with master or an issue with your patch on top of master?

> 
> There are a few other fixes that ended up in here. Since code is currently churning on this repo, I hope you don't mind them getting lumped into this. I'll try to leave comments on a few points of the PR that might need to be discussed further.

Ideally unrelated changes would be split into different commits, but I'm not too picky here. Thanks for the comments

> 
> Please feel free to make any changes.
> 
> EDIT: By 'near-complete' I mean I haven't been able to fully test the end-to-end workflow of initializing a disk, writing a file into it, and loading it on XEMU/HW with master regressed. Also I imagine there might be some questions/changes/cleanup that may be requested +1

I'd really like to see some tests added that do some formatting. We can probably ditch the existing drive image and just format a blank disk image now

gaasedelen commented:
Thanks for all the comments. I agree with all of them and will try to get them sorted out today (or in the coming days).

> It will not be exactly 1:1 with a retail box out of the factory because it is missing the config partition.

True, actually. I forgot about the refurb/config sectors. I don't recall them having much but I'll loop back and take a look at initializing any relevant data in them.

> Weird. The tester script writes a (~64K) file and reads it back out. You have confirmed it's an issue with master or an issue with your patch on top of master?

I will double check. 

Last night when I noticed writing into the disk was failing on my branch I did a checkout of master and re-ran cmake & make clean & make. I tried writing `xboxdash.xbe` into the C drive of a few different clean drive images I had and was getting an I/O error at 8192 bytes.

Checking out the commit (44a03f82b18f4a49a6a4b02361cc469718e7af76) I had been using for inserting files onto 4K sector disks previously worked okay, including images initialized by the submitted PR. It's possible I had some stale artifacts leftover from my branch when building master, or maybe I just fat-fingered some commands as I was pretty tired. 

That said, I would not be surprised if my branch broke writes in a different way (eg, the `FATX_CLUSTER_DATA` seems glaring lol). I'll get things sorted out, either way. I just wanted to give a heads up and get the PR on git in-case I didn't have time to troubleshoot this week.

gaasedelen commented:
Just wanted to confirm that the 'writing files' breakage was indeed my bad, the primary issue had to do with the `FATX_CLUSTER_DATA` stuff you called out in the switch/case rewrite of `fatx_get_fat_entry_type(...)` which I have further simplified.

I also noticed that after rebasing onto master `fatxfs` was not building by default on my machine anymore, so yes I must have been using stale artifacts in addition to the breakages. The problem is that `FUSE_MODULE_NAME` is not being defined by the `pkg` lookup stuff in the CMake file, but based on [this](https://cmake.org/cmake/help/v3.0/module/FindPkgConfig.html) I swapped in `FUSE_FOUND` which seems to work.

I am building on Ubuntu 20.04 with CMake v3.10.2 for reference

gaasedelen commented:
Also just wanted to confirm that with the above commit I was able to take a blank 16gb file, format it as 4K retail, copy evox onto it as `xboxdash.xbe`, and boot it under XEMU.

I will still go through and address the remaining feedback points and bring any other final thoughts up when I'm done.

GXTX commented:
> It will not be exactly 1:1 with a retail box out of the factory because it is missing the config partition

Config sector at 0x600 starts with
```
type configHeader struct {
	Magic [4]byte // BRFR
	NumberOfBoots uint32
	FormatDate [8]byte // FILETIME
}
```

It's possible that network settings would have to be set as well for dashes that used post-live as the live upgrader would correctly initialize these values. With broken network settings I don't believe live functions on retail would work as it'd just see a corrupted 'sector' at 0x1000 (live is dead so it's a non-issue, but for completeness..).

gaasedelen commented:
Resolved the bulk of the comments, here's a few outstanding items:

- [ ] Simplify superblock init (move read OR generate decision into dedicated superblock init func, should be easy)
- [ ] Create a default mechanism to 'mount' F partition
    - Add as a 'non-retail' entry to the existing partition table?
    - Have to define behavior for 'takes-all' size resolution
 - [ ] Initialize config sectors
 - [ ] Generate disk without requiring existing device or file!!?!?
 
I only have `retail` and `f-takes-all` for now because I think those are the most important. 

I imagine we can add a flag to 'manually' initialize a single partition to spec a bit later on. I also really don't want to encourage formatting new disks with 'F & G' or '137gb' (LBA28) as I'd like to consider them legacy formats.

GXTX commented on src/fatx_internal.h:
```diff
@@ -75,7 +81,18 @@
+struct fatx_refurb_info {
+    uint32_t signature;
+    uint32_t number_of_boots;
+    uint64_t first_power_on;
```
`first_power_on` isn't set by the system, it's set during manufacturing as a format date.

gaasedelen replied:
Tbh I don't have a untouched box from the factory to counter this (nor do I think it matters) -- I just picked the name based on the fact that the dash initializes the timestamp if it isn't already set:

![image](https://user-images.githubusercontent.com/9522648/154183261-0fafa438-73d9-4682-b033-334222ba8463.png)

This is `xboxdash.xbe` I pulled off my retail and applied FLIRT + some additional cleanup to ^

GXTX replied:
til

gaasedelen commented:
So I think this is mostly ready to go. I'll try to do a little more testing tomorrow or address any final comments you might have in order to get this merged @mborgerson.

I think that generating a disk without an existing file/device should be made as a separate PR. It's a useful feature and one that should have careful attention paid to it, but could also snowball in terms of scope for ease-of-use / configurability / usability: 
```
fatxfs -retail
fatxfs --generate=raw --format=retail
fatxfs --generate=qcow2 --format=f-takes-all --disk-size=16TB --sector-size=4096
```
I was most interested in being able to initialize HW devices over USB with this PR, and I feel that I've achieved that.

Thanks again for taking the time to review and consider this work.

mborgerson commented:
> So I think this is mostly ready to go. I'll try to do a little more testing tomorrow or address any final comments you might have in order to get this merged @mborgerson.

Awesome, I'll take a look at the updates

> I think that generating a disk without an existing file/device should be made as a separate PR. It's a useful feature and one that should have careful attention paid to it, but could also snowball in terms of scope for ease-of-use / configurability / usability:

It's easy enough to create a blank file (analogous to plugging in a real device) first and initialize the filesystems in it so I'm definitely okay with deferring it.

> I was most interested in being able to initialize HW devices over USB with this PR, and I feel that I've achieved that. So I think it'll have to be someone else who contributes pure disk generation :-P.

Works for me

> Thanks again for taking the time to review and consider this work.

Thank you for the contribution!

mborgerson commented on src/fatx.h:
```diff
@@ -42,6 +43,11 @@
+#define FATX_RETAIL_CLUSTER_SIZE     (16 * 1024)
+#define FATX_RETAIL_PARTITION_COUNT  5
+
+#define FATX_READ_FROM_SUPERBLOCK    0
```
I think it's not exactly clear that this should be specified in `fatx_open_device`'s `sectors_per_cluster` argument, and only in that argument. Can you add a comment here?

mborgerson commented on src/fatx_disk.c:
```diff
@@ -0,0 +1,294 @@
+    struct fatx_partition_map_entry const *pi;
+    size_t f_offset, f_size;
+
+    if (format_type == FATX_FORMAT_NONE)
```
It's not immediately clear what format scheme of `NONE` means. It could be interpreted as blanking the partition, for instance. I think perhaps a better name would be `FATX_FORMAT_INVALID`? And in this case `fatx_disk_format` should return an error

mborgerson commented on src/fatxfs_fuse.c:
```diff
@@ -617,19 +610,24 @@ void fatx_fuse_print_usage(void)
+                    "    --loglevel=<level>             control the log output level (a higher value yields more output)\n\n"
+                    "Disk formatting options:\n"
+                    "    --format=<format>              specify the format (retail, f-takes-all) to initialize the device to\n"
+                    "    --sector-size=<size>           specify the size (in bytes) of a device sector (default is 512)\n"
```
This is already specified above, don't need to duplicate it

mborgerson commented on src/fatxfs_fuse.c:
```diff
@@ -617,19 +610,24 @@ void fatx_fuse_print_usage(void)
+                    "    --log=<log path>               enable fatxfs logging\n"
+                    "    --loglevel=<level>             control the log output level (a higher value yields more output)\n\n"
+                    "Disk formatting options:\n"
+                    "    --format=<format>              specify the format (retail, f-takes-all) to initialize the device to\n"
```
It would be nice to support formatting XMU memory card partitions, which take the whole disk, use a sector size of 4096, and sectors per cluster of 4. But we can defer it to a follow up

mborgerson commented on src/fatx_disk.c:
```diff
@@ -0,0 +1,294 @@
+/*
+ * Read refurb sector.
+ */
+int fatx_disk_read_refurb_info(char const *path)
```
This doesn't give caller the refurb info, but it's also not used anywhere so I think we can just remove it for now

gaasedelen commented on src/fatx_disk.c:
```diff
@@ -0,0 +1,294 @@
+{
+    struct fatx_partition_map_entry const *pi;
+
+    for (int i = 0; i < sizeof(fatx_partition_map); i++)
```
big ooooof. good catch, was moving a bit too fast there.

gaasedelen commented on src/fatx_internal.h:
```diff
@@ -74,26 +70,41 @@
 
 /* FAT entry types (not the actual value of the entry). */
 #define FATX_CLUSTER_AVAILABLE       0
-#define FATX_CLUSTER_RESERVED        1
```
Technically I don't think RESERVED (0xFFFFFFF0) should ever show up in the actual FAT table as an entry, I think it's meant more to serve as a clear-cut boundary to developers above which they are free to define special entry ID's (BAD, MEDIA, LAST).

That said, I went back and re-added it (... which also helped me catch a small edgecase I missed)

mborgerson replied:
Oh nice!

gaasedelen commented:
I addressed all your comments and also just tried to push a small fix for the python binding stuff (?)

Anyway, I'll do some final testing tonight on both Hardware and XEMU to make sure everything seems good to go.

mborgerson commented:
> I addressed all your comments and also just tried to push a small fix for the python binding stuff (?)
> 
> Anyway, I'll do some final testing tonight on both Hardware and XEMU to make sure everything seems good to go.

@gaasedelen Would you be willing to fix the failing CI? I think it'll be good to merge once CI passes and you're done testing whatever you wanted to test

gaasedelen commented:
I was able to do some additional testing and there's one final 'minor' change that I'd like to slip into this PR if possible. I'll try to touch base about it later today.

I'll also take a look at the remaining CI failure.

gaasedelen commented:
So the last change I'm slipping in again, might be a little controversial but it's in support of extended disks (>2TB) on real hardware. 

It can take several minutes for hardware to mount a large FATX volume formatted with 64KB clusters. This is due to the sheer physical size of the FAT (potentially, gigabytes) which needs to scanned by the CPU when mounting the volume.

Boosting clusters to 512kb indeed, can be 'more wasteful' of space but has several benefits including dramatically faster mounting (~120mb for a 16TB disk versus gigabytes). I think we'll be seeing this added to other FATX tools and frequently used in the coming era of 'super-large' Xbox disk support.

Also, the CI issue should hopefully be fixed.

mborgerson commented:
> So the last change I'm slipping in again, might be a little controversial but it's in support of extended disks (>2TB) on real hardware.
> 
> It can take several minutes for hardware to mount a large FATX volume formatted with 64KB clusters. This is due to the sheer physical size of the FAT (potentially, gigabytes) which needs to scanned by the CPU when mounting the volume.
> 
> Boosting clusters to 512kb indeed, can be 'more wasteful' of space but has several benefits including dramatically faster mounting (~120mb for a 16TB disk versus gigabytes). I think we'll be seeing this added to other FATX tools and frequently used in the coming era of 'super-large' Xbox disk support.
> 
> Also, the CI issue should hopefully be fixed.
> 
> Edit: formatting with 'larger clusters' would be through the `--sectors-per-cluster` command that was previously added by this PR. This boosts it from a max of 128 'sectors per cluster' to 1024.

I like it! I don't think 512kb is an unreasonable cluster size anyway. In the future we should choose an appropriate cluster size based on the disk size being formatted automatically.

Thanks for the patches!
